### PR TITLE
AF-2397 :Deletion of group from Business Central UI does not remove group from the security-policy.properties file

### PR DESCRIPTION
**Pull request by RishiRajAnand**
-  Added feature to remove a specific group from security-policy.properties
-  If the user adds a group with the same name as deleted group, default configurations will be applied to the newly added group.

### CODE REVIEW

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins retest this please

**Comment by RishiRajAnand** [comment]
Jenkins retest this please

**Comment by RishiRajAnand** [comment]
Jenkins retest this please

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by ederign** [comment]
@caponetto @barboras7 can you please take a look at this?

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java [hunk 1]
```diff
@@ -149,7 +150,34 @@ public FileVisitResult visitFile(final Path file,
...
-        return fileName.equals("security-policy.properties") || fileName.startsWith("security-module-");
+        return fileName.equals(POLICY_FILE) || fileName.startsWith("security-module-");
+    }
+
+    @Override
+    public void deletePolicyByGroup(Group group, AuthorizationPolicy policy) {
+        ioService.startBatch(fileSystem,
+                             new CommentedOption("system",
+                                                 "Delete Policy"));
+        try (StringWriter sw = new StringWriter()){
```
Comment (caponetto) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java]: I see that lines 161 to 174 have almost the same code as the `savePolicyIntoVfs` method.
Consider extracting duplicated code into generic methods.

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java [hunk 2]
```diff
@@ -323,6 +323,39 @@ public void write(Group group,
         }
     }
 
+    public void remove (Group group , AuthorizationPolicy policy,
```
Comment (caponetto) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java]:
```suggestion
    public void remove(Group group, AuthorizationPolicy policy,
```

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java [hunk 3]
```diff
@@ -323,6 +323,39 @@ public void write(Group group,
...
+                       out);
+                remove(subject,
+                      policy.getPermissions(subject),
+                      out);
+            }
+        }
+    }
+
+    public void remove(Group group,
+                      Map out) {
```
Comment (caponetto) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java]:
```suggestion
                       Map out) {
```

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java [hunk 4]
```diff
@@ -323,6 +323,39 @@ public void write(Group group,
...
+                        Map out) {
+
+        write(policy,out);
+
+        for (Group subject : policy.getGroups()) {
+            if (group.getName().equals(subject.getName())){
+                remove(subject,
+                       out);
+                remove(subject,
+                      policy.getPermissions(subject),
```
Comment (caponetto) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java]:
```suggestion
                       policy.getPermissions(subject),
```

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java [hunk 5]
```diff
@@ -323,6 +323,39 @@ public void write(Group group,
...
+                      out);
```
Comment (caponetto) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java]:
```suggestion
                       out);
```

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationServiceImpl.java [hunk 6]
```diff
@@ -50,4 +51,12 @@ public void savePolicy(AuthorizationPolicy policy) {
         permissionManager.setAuthorizationPolicy(policy);
         savedEvent.fire(new AuthorizationPolicySavedEvent(policy));
     }
+
+    @Override
+    public void deletePolicyByGroup(Group group, AuthorizationPolicy policy) {
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationServiceImpl.java]: added

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java [hunk 7]
```diff
@@ -149,7 +150,34 @@ public FileVisitResult visitFile(final Path file,
 
     public boolean isPolicyFile(Path p) {
         String fileName = p.getName(p.getNameCount() - 1).toString();
-        return fileName.equals("security-policy.properties") || fileName.startsWith("security-module-");
+        return fileName.equals(POLICY_FILE) || fileName.startsWith("security-module-");
+    }
+
+    @Override
+    public void deletePolicyByGroup(Group group, AuthorizationPolicy policy) {
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java]: added

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java (hunk 1 above)
Comment (RishiRajAnand) [reply uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java]: removed duplicated code

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by caponetto** [comment]
Hi @RishiRajAnand, could you take a look at the code smells raised by SonarCloud?
They are easy to get rid of.

**Comment by RishiRajAnand** [comment]
> Hi @RishiRajAnand, could you take a look at the code smells raised by SonarCloud?
> They are easy to get rid of.

Sure, Thanks

**Comment by barboras7** [comment]
jenkins execute full downstream build

File: uberfire-backend/uberfire-backend-server/src/test/java/org/uberfire/backend/server/authz/AuthzPolicyDeployerTest.java [hunk 8]
```diff
@@ -95,6 +105,42 @@ public void testPolicyLoad2() throws Exception {
...
+        verify(permissionManager).setAuthorizationPolicy(policy);
+
+        Set<Group> roles = policyCaptor.getValue().getGroups();
+        assertEquals(1, roles.size());
+
+        PermissionCollection permissions = policy.getPermissions(group);
+        Permission permission = permissions.get("perspective.read");
+        assertNotNull(permission);
+        assertEquals(AuthorizationResult.ACCESS_GRANTED, permission.getResult());
+        ;
```
Comment (barboras7) [inline uberfire-backend/uberfire-backend-server/src/test/java/org/uberfire/backend/server/authz/AuthzPolicyDeployerTest.java]: unnecessary semicolon

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/group/workflow/GroupEditorWorkflow.java [hunk 9]
```diff
@@ -163,14 +164,26 @@ public boolean isDirty() {
...
+        AuthorizationPolicy authzPolicy = permissionManager.getAuthorizationPolicy();
+        showLoadingBox();
+        authorizationService.call(r -> {
+                                      hideLoadingBox();
+                                      deleteGroupEvent.fire(new DeleteGroupEvent(name));
+                                      workbenchNotification.fire(new NotificationEvent(UsersManagementWidgetsConstants.INSTANCE.groupRemoved(name),
+                                                                                       INFO));
+                                  },
+                                  errorCallback).deletePolicyByGroup(group, authzPolicy);
+
```
Comment (barboras7) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/group/workflow/GroupEditorWorkflow.java]: empty line

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java [hunk 10]
```diff
@@ -149,10 +157,22 @@ public FileVisitResult visitFile(final Path file,
...
     public boolean isPolicyFile(Path p) {
         String fileName = p.getName(p.getNameCount() - 1).toString();
-        return fileName.equals("security-policy.properties") || fileName.startsWith("security-module-");
+        return fileName.equals(POLICY_FILE) || fileName.startsWith("security-module-");
     }
 
-    public void savePolicyIntoVfs(AuthorizationPolicy policy,
+    @Override
+    public void deletePolicyByGroup(Group group, AuthorizationPolicy policy) {
+
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyVfsStorage.java]: removed empty line

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java [hunk 11]
```diff
@@ -323,6 +323,39 @@ public void write(Group group,
...
+                remove(subject,
+                       out);
+                remove(subject,
+                       policy.getPermissions(subject),
+                       out);
+            }
+        }
+    }
+
+    public void remove(Group group,
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/authz/AuthorizationPolicyMarshaller.java]: These methods are not used directly by other classes at the moment, so it makes sense to make this private. However, these methods are generic and can be used for specific use cases in future.

**Comment by barboras7** [comment]
jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
jenkins execute full downstream build

**Comment by ederign** [comment]
jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
jenkins execute full downstream build

**Pull request merged**